File: PeerConnect F/src/services/Chatservice.js
import apiService from './api.js';

class ChatService {
  // Send message to project chat - matches ChatController endpoint
  async sendMessage(projectId, content) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }
      if (!content || content.trim().length === 0) {
        throw new Error('Message content is required');
      }

      console.log('ChatService: Sending message to project:', projectId);
      const response = await apiService.post(`/chat/projects/${projectId}/messages`, {
        content: content.trim()
      });
      console.log('ChatService: Message sent successfully:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error sending message:', error);
      throw new Error(error.message || 'Failed to send message');
    }
  }

  // Get project messages with pagination - matches ChatController endpoint
  async getProjectMessages(projectId, page = 0, size = 50) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }

      console.log('ChatService: Getting messages for project:', projectId);
      const params = { page, size };
      const response = await apiService.get(`/chat/projects/${projectId}/messages`, params);
      console.log('ChatService: Messages retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting project messages:', error);
      throw new Error(error.message || 'Failed to get project messages');
    }
  }

  // Get recent messages (last 24 hours) - matches ChatController endpoint
  async getRecentMessages(projectId) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }

      console.log('ChatService: Getting recent messages for project:', projectId);
      const response = await apiService.get(`/chat/projects/${projectId}/messages/recent`);
      console.log('ChatService: Recent messages retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting recent messages:', error);
      throw new Error(error.message || 'Failed to get recent messages');
    }
  }

  // Search messages - matches ChatController endpoint
  async searchMessages(projectId, query, page = 0, size = 20) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }
      if (!query || query.trim().length === 0) {
        throw new Error('Search query is required');
      }

      console.log('ChatService: Searching messages in project:', projectId, 'with query:', query);
      const params = { query: query.trim(), page, size };
      const response = await apiService.get(`/chat/projects/${projectId}/messages/search`, params);
      console.log('ChatService: Search results:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error searching messages:', error);
      throw new Error(error.message || 'Failed to search messages');
    }
  }

  // Get messages after specific time - matches ChatController endpoint
  async getMessagesAfter(projectId, afterTime) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }
      if (!afterTime) {
        throw new Error('After time is required');
      }

      console.log('ChatService: Getting messages after:', afterTime);
      const params = { after: afterTime };
      const response = await apiService.get(`/chat/projects/${projectId}/messages/after`, params);
      console.log('ChatService: Messages after time retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting messages after time:', error);
      throw new Error(error.message || 'Failed to get messages after specified time');
    }
  }

  // Edit message - matches ChatController endpoint
  async editMessage(messageId, content) {
    try {
      if (!messageId) {
        throw new Error('Message ID is required');
      }
      if (!content || content.trim().length === 0) {
        throw new Error('Message content is required');
      }

      console.log('ChatService: Editing message:', messageId);
      const response = await apiService.put(`/chat/messages/${messageId}`, {
        content: content.trim()
      });
      console.log('ChatService: Message edited successfully:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error editing message:', error);
      throw new Error(error.message || 'Failed to edit message');
    }
  }

  // Delete message - matches ChatController endpoint
  async deleteMessage(messageId) {
    try {
      if (!messageId) {
        throw new Error('Message ID is required');
      }

      console.log('ChatService: Deleting message:', messageId);
      const response = await apiService.delete(`/chat/messages/${messageId}`);
      console.log('ChatService: Message deleted successfully:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error deleting message:', error);
      throw new Error(error.message || 'Failed to delete message');
    }
  }

  // Get latest message in project - matches ChatController endpoint
  async getLatestMessage(projectId) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }

      console.log('ChatService: Getting latest message for project:', projectId);
      const response = await apiService.get(`/chat/projects/${projectId}/messages/latest`);
      console.log('ChatService: Latest message retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting latest message:', error);
      throw new Error(error.message || 'Failed to get latest message');
    }
  }

  // Get message count for project - matches ChatController endpoint
  async getMessageCount(projectId) {
    try {
      if (!projectId) {
        throw new Error('Project ID is required');
      }

      console.log('ChatService: Getting message count for project:', projectId);
      const response = await apiService.get(`/chat/projects/${projectId}/messages/count`);
      console.log('ChatService: Message count retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting message count:', error);
      throw new Error(error.message || 'Failed to get message count');
    }
  }

  // Get message by ID - matches ChatController endpoint
  async getMessageById(messageId) {
    try {
      if (!messageId) {
        throw new Error('Message ID is required');
      }

      console.log('ChatService: Getting message by ID:', messageId);
      const response = await apiService.get(`/chat/messages/${messageId}`);
      console.log('ChatService: Message retrieved:', response);
      return response;
    } catch (error) {
      console.error('ChatService: Error getting message by ID:', error);
      throw new Error(error.message || 'Failed to get message');
    }
  }

  // Real-time polling helper (since WebSocket isn't implemented)
  async pollForNewMessages(projectId, lastMessageTime, intervalMs = 5000) {
    try {
      console.log('ChatService: Starting message polling for project:', projectId);
      
      const pollMessages = async () => {
        try {
          const messages = await this.getMessagesAfter(projectId, lastMessageTime);
          return messages;
        } catch (error) {
          console.error('ChatService: Polling error:', error);
          return null;
        }
      };

      return setInterval(pollMessages, intervalMs);
    } catch (error) {
      console.error('ChatService: Error setting up polling:', error);
      throw new Error('Failed to set up message polling');
    }
  }

  // Stop polling
  stopPolling(intervalId) {
    if (intervalId) {
      clearInterval(intervalId);
      console.log('ChatService: Message polling stopped');
    }
  }
}

export const chatService = new ChatService();
export default chatService;

File: QUASAR-backend/src/main/java/com/ADP/peerConnect/service/Impl/ChatService.java
package com.ADP.peerConnect.service.Impl;

import com.ADP.peerConnect.exception.BadRequestException;
import com.ADP.peerConnect.exception.ResourceNotFoundException;
import com.ADP.peerConnect.exception.UnauthorizedException;
import com.ADP.peerConnect.model.entity.ChatMessage;
import com.ADP.peerConnect.model.entity.Project;
import com.ADP.peerConnect.model.entity.User;
import com.ADP.peerConnect.model.enums.MessageType;
import com.ADP.peerConnect.repository.ChatMessageRepository;
import com.ADP.peerConnect.service.Interface.iChatService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service class for Chat message operations
 */
@Service
@Transactional
public class ChatService implements iChatService {
    
    @Autowired
    private ChatMessageRepository chatMessageRepository;
    
    @Autowired
    private ProjectService projectService;
    
    @Autowired
    private UserService userService;
    
    /**
     * Send message to project chat
     */
    public ChatMessage sendMessage(String projectId, String senderId, String content) {
        // Validate input
        if (content == null || content.trim().isEmpty()) {
            throw new BadRequestException("Message content cannot be empty");
        }
        
        // Check if project exists
        Project project = projectService.findById(projectId);
        
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, senderId)) {
            throw new UnauthorizedException("You must be a project member to send messages");
        }
        
        // Get sender
        User sender = userService.findById(senderId);
        
        // Create message
        ChatMessage message = new ChatMessage();
        message.setProject(project);
        message.setSender(sender);
        message.setMessage(content.trim());
        message.setMessageType(MessageType.TEXT);
        message.setCreatedAt(LocalDateTime.now());
        
        return chatMessageRepository.save(message);
    }
    
    /**
     * Edit message
     */
    public ChatMessage editMessage(Long messageId, String userId, String newContent) {
        // Validate input
        if (newContent == null || newContent.trim().isEmpty()) {
            throw new BadRequestException("Message content cannot be empty");
        }
        
        ChatMessage message = findById(messageId);
        
        // Check if user is the sender
        if (!message.getSender().getId().equals(userId)) {
            throw new UnauthorizedException("You can only edit your own messages");
        }
        
        // Update message
        message.setMessage(newContent.trim());
        
        return chatMessageRepository.save(message);
    }
    
    /**
     * Delete message
     */
    public void deleteMessage(Long messageId, String userId) {
        ChatMessage message = findById(messageId);
        
        // Check if user is the sender or project Lead
        boolean isSender = message.getSender().getId().equals(userId);
        boolean isProjectLead = projectService.isProjectLead(message.getProject().getId(), userId);
        
        if (!isSender && !isProjectLead) {
            throw new UnauthorizedException("You can only delete your own messages or if you're the project Lead");
        }
        
        chatMessageRepository.delete(message);
    }
    
    /**
     * Get project messages
     */
    public Page<ChatMessage> getProjectMessages(String projectId, String userId, Pageable pageable) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to view messages");
        }
        
        return chatMessageRepository.findByProjectIdOrderByCreatedAtDesc(projectId, pageable);
    }
    
    /**
     * Get recent messages (last 24 hours)
     */
    public List<ChatMessage> getRecentMessages(String projectId, String userId) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to view messages");
        }
        
        LocalDateTime since = LocalDateTime.now().minusHours(24);
        return chatMessageRepository.findRecentMessages(projectId, since);
    }
    
    /**
     * Search messages
     */
    public Page<ChatMessage> searchMessages(String projectId, String query, String userId, Pageable pageable) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to search messages");
        }
        
        return chatMessageRepository.searchMessagesByContent(projectId, query, pageable);
    }
    
    /**
     * Get messages after a specific time (for real-time updates)
     */
    public List<ChatMessage> getMessagesAfter(String projectId, LocalDateTime after, String userId) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to view messages");
        }
        
        return chatMessageRepository.findMessagesAfter(projectId, after);
    }
    
    /**
     * Get latest message in project
     */
    public ChatMessage getLatestMessage(String projectId, String userId) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to view messages");
        }
        
        List<ChatMessage> messages = chatMessageRepository.findLatestMessage(projectId, PageRequest.of(0, 1));
        return messages.isEmpty() ? null : messages.get(0);
    }
    
    /**
     * Get message count for project
     */
    public long getMessageCount(String projectId, String userId) {
        // Check if user is a member of the project
        if (!projectService.isProjectMember(projectId, userId)) {
            throw new UnauthorizedException("You must be a project member to view message count");
        }
        
        return chatMessageRepository.countByProjectId(projectId);
    }
    
    /**
     * Send system message
     */
    public ChatMessage sendSystemMessage(String projectId, String content) {
        Project project = projectService.findById(projectId);
        
        ChatMessage message = new ChatMessage();
        message.setProject(project);
        message.setSender(null); // System messages have no sender
        message.setMessage(content);
        message.setMessageType(MessageType.SYSTEM);
        message.setCreatedAt(LocalDateTime.now());
        
        return chatMessageRepository.save(message);
    }
    
    /**
     * Find message by ID
     */
    public ChatMessage findById(Long messageId) {
        return chatMessageRepository.findById(messageId)
            .orElseThrow(() -> new ResourceNotFoundException("Message not found"));
    }
    
    /**
     * Get user's messages across all projects
     */
    public Page<ChatMessage> getUserMessages(String userId, Pageable pageable) {
        return chatMessageRepository.findBySenderIdOrderByCreatedAtDesc(userId, pageable);
    }
    
    /**
     * Get messages from multiple projects (for user's project list)
     */
    public Page<ChatMessage> getMessagesFromProjects(List<String> projectIds, Pageable pageable) {
        return chatMessageRepository.findByProjectIds(projectIds, pageable);
    }
}
